contracts/src/Gateway.sol

```diff
@@ -578,43 +578,47 @@ contract Gateway is IGateway, IInitializable {
             revert Unauthorized();
         }
 
-        Config memory config = abi.decode(data, (Config));
-
         CoreStorage.Layout storage core = CoreStorage.layout();
-        core.mode = config.mode;
 
-        // Initialize agent for BridgeHub
-        address bridgeHubAgent = address(new Agent(BRIDGE_HUB_AGENT_ID));
-        core.agents[BRIDGE_HUB_AGENT_ID] = bridgeHubAgent;
+        // Check if initialized and run initialization once.
+        if (core.channels[PRIMARY_GOVERNANCE_CHANNEL_ID].agent == address(0)) {
```

Comment:
This quite an invasive way of implementing this check, introducing a lot of code churn. 

Rather invert the check:

```solidity
if (core.channels[PRIMARY_GOVERNANCE_CHANNEL_ID].agent != address(0)) {
  revert AlreadyInitialized()
}
```

As the above example shows, I think we should probably also fail the upgrade by reverting.